File: p309m/max_profit_cooldown.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function (prices) {
  let [dpClosed, dpClosedPrev, dpOpen] = [0, 0, Number.MIN_SAFE_INTEGER];
  for (let p of prices) {
    let oldClosed = dpClosed;
    dpClosed = Math.max(dpClosed, dpOpen + p); // hold or sell
    dpOpen = Math.max(dpOpen, dpClosedPrev - p); // hold or buy
    dpClosedPrev = oldClosed;
  }
  return dpClosed;
};
// TEST
[
  [[2], 0],
  [[1, 3], 2],
  [[1, 2, 3, 0, 2], 3],
  [[1, 3, 2, 8, 4, 9], 8],
].forEach((t) => {
  const actual = maxProfit(t[0]);
  console.log(
    'Max profit you can make from',
    t[0],
    'with cooldown',
    '->',
    actual
  );
  console.assert(actual === t[1]);
});

File: p309m/max_profit_cooldown.py
from typing import List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        dp_i0, dp_i0_prev, dp_i1 = 0, 0, float("-Inf")
        for p in prices:
            i0_old = dp_i0
            dp_i0 = max(dp_i0, dp_i1 + p)
            dp_i1 = max(dp_i1, dp_i0_prev - p)
            dp_i0_prev = i0_old
        return dp_i0


# TESTS
tests = [
    ([2], 0),
    ([1, 3], 2),
    ([1, 2, 3, 0, 2], 3),
    ([1, 3, 2, 8, 4, 9], 8),
]
for t in tests:
    sol = Solution()
    actual = sol.maxProfit(t[0])
    print("Max profit you can make from", t[0], "with cooldown ->", actual)
    assert actual == t[1]

